### PR TITLE
RavenDB-24112 Handle correct ID with non-ascii character and escaped characters

### PR DESCRIPTION
**Pull request by haludi**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-24112

Previously, when the ID included non-ASCII characters, FindEscapePositions was called on a temporary buffer. During this step, _jsonParserState collected the escape positions and also modified the ID by escaping characters with values lower than 32.

However, WriteEscapePositionsTo was later called on a different copy of the ID that did not include these modifications. As a result, characters that were escaped into \uXXXX increased the size of the ID, and the previously calculated escape positions no longer matched the final content.

I’m not sure what the motivation was for using the temporary buffer in this step. I compared this logic to the case where the ID doesn’t contain non-ASCII characters, where we simply call FindEscapePositionsIn directly on the final ID.

The underlying issue seems to be that FindEscapePositionsIn both finds escape positions and transforms the input by escaping control characters (ASCII < 32). While that transformation is sometimes the desired outcome, the method name doesn’t make this clear — which can cause confusion in its usage and lead to bugs like the one addressed in this PR.

### Type of change
- [x] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [ ] New feature

### How risky is the change?
- [ ] Low 
- [ ] Moderate 
- [x] High
- [ ] Not relevant

### Backward compatibility
- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?
- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update
- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor
- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing
- [ ] Existing tests verify the correct behavior

### Testing by RavenDB QA team
- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?
- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work
- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/DocumentIdWorker.cs [hunk 1]
```diff
@@ -378,23 +380,25 @@ private static ByteStringContext.InternalScope UnicodeGetLowerIdAndStorageKey<TT
...
                 var actualIdLenSize = JsonParserState.VariableSizeIntSize(idSize);
                 if (actualIdLenSize < maxIdLenSize)
                 {
                     var movePtr = maxIdLenSize - actualIdLenSize;
                     id += movePtr;
-                    writePos += movePtr;
                 }
 
+                byte* writePos = id;
+                _jsonParserState.FindEscapePositionsAndEscapeControls(id + maxIdLenSize, ref idSize, escapePositionsSize);
```
Comment (haludi) [inline src/Raven.Server/Documents/DocumentIdWorker.cs]: Main fix:
This change performs the escape operation directly on the final ID, ensuring that control characters (ASCII < 32) are properly escaped as part of the process.

**Review by grisha-kotler — APPROVED** [review]

**Comment by haludi** [comment]
I wanted to test an edge case where the ID is made entirely of repeated \r characters.
In this case, RavenDB treats it as if no ID was provided and generates one on the server side.
The check is done using `string.IsNullOrWhiteSpace`:
https://github.com/ravendb/ravendb/blob/release/v5.4/src/Raven.Server/Documents/DocumentPutAction.cs#L410

Is this intentional, or should it be `string.IsNullOrEmpty` instead?

This might not be a common scenario, but since RavenDB allows a lot of freedom with IDs, this limitation feels a bit unexpected.
Someone might want to create a document per character and use that character as the ID.
This can, of course, be worked around by using a collection prefix like Characters/\r.

**Review by grisha-kotler — APPROVED** [review]

**Comment by ayende** [comment]
The white space check is intentional, to avoid `"    "`  being a valid id, etc.

File: src/Sparrow/Json/Parsing/JsonParserState.cs [hunk 2]
```diff
@@ -131,18 +130,18 @@ public static int FindEscapePositionsMaxSize(byte* str, int size, out int escape
...
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public void FindEscapePositionsIn(byte* str, ref int len, int previousComputedMaxSize)
+        public void FindEscapedPositionsAndEscapeControls(byte* str, ref int len, int previousComputedMaxSize)
         {
-            FindEscapePositionsIn(EscapePositions, str, ref len, previousComputedMaxSize);
+            FindEscapedPositionsAndEscapeControls(EscapePositions, str, ref len, previousComputedMaxSize);
         }
 
-        public static void FindEscapePositionsIn(FastList<int> buffer, byte* str, ref int len, int previousComputedMaxSize)
+        public static void FindEscapedPositionsAndEscapeControls(FastList<int> buffer, byte* str, ref int len, int previousComputedMaxSize)
```
Comment (arekpalinski) [inline src/Sparrow/Json/Parsing/JsonParserState.cs:138]: The motivation for the rename was to express that this method does 2 things (what is actually quite surprising, especially when it was named "FindEscapePositionsIn"):

- it finds escape positions
- it escapes the control characters

If anyone is able to suggest something better, please feel free to do it.

**Review by arekpalinski — APPROVED** [review]
I reviewed this PR with @haludi during a call, where he guided me through and explained everything in detail.

We also discussed backward compatibility. This change only affects how we _write_ data, so there are no concerns regarding reading. Initially, I was worried that the change might cause _other / new_ issues when attempting to read existing corrupted data (e.g AVE because we assume data to be formatted differently), but we confirmed that such cases are not possible.
